Minor improvements to ec2 service discovery

federicobaldo opened the pull request:
Hello,
I've tried using this new feature although for our environment I had to change the following: 

1. static credentials replaced with defaults.DefaultChainCredentials.This change ensures that credentials are sourced from all possible providers available with the aws sdk, in the following order: env variables, shared awsconfig file in user folder, ec2 instance role.
[https://docs.aws.amazon.com/sdk-for-go/api/aws/defaults.html](https://docs.aws.amazon.com/sdk-for-go/api/aws/defaults.html)

2. if the ec2 instace has a public ip address, set the address label to the public dns name of the instance. This play nicely with ec2 split-horizon dns (the private ip is resolved within the same ec2 region) and various security group rules which relies on private ips (e.g. allowing incoming connections from another security group).

3. Added a few labels: AvailabilityZone, PublicDns, VpcId (if available), SubnetId (if in Vpc)

I think that those may be helpful for  others as well.

Sending this PR to @mjibson and @fabxc as they already worked on this.

Thanks,
Federico

fabxc commented on retrieval/discovery/ec2.go:
```diff
@@ -29,10 +30,14 @@ import (
 
 const (
 	ec2Label           = model.MetaLabelPrefix + "ec2_"
+	ec2LabelAz         = ec2Label + "az"
```
Please capitalize all acronyms for consistency.

fabxc commented:
Thanks! This all sounds reasonable to me.
Of course input from @mjibson would be helpful here, too.

mjibson commented on retrieval/discovery/ec2.go:
```diff
@@ -29,10 +30,14 @@ import (
 	ec2Label           = model.MetaLabelPrefix + "ec2_"
+	ec2LabelAz         = ec2Label + "az"
 	ec2LabelInstanceID = ec2Label + "instance_id"
+	ec2LabelPublicDns  = ec2Label + "public_dns_name"
```
Dns -> DNS

mjibson commented on retrieval/discovery/ec2.go:
```diff
@@ -29,10 +30,14 @@ import (
+	ec2LabelPublicDns  = ec2Label + "public_dns_name"
 	ec2LabelPublicIP   = ec2Label + "public_ip"
 	ec2LabelPrivateIP  = ec2Label + "private_ip"
+	ec2LabelSubnetId   = ec2Label + "subnet_id"
```
Id -> ID

brian-brazil replied:
There's 35 fields currently in the output of this rpc, and more will be added over time. Can we do this in a way that doesn't require us to maintain individual mappings?

mjibson commented on retrieval/discovery/ec2.go:
```diff
@@ -29,10 +30,14 @@ import (
 	ec2LabelPrivateIP  = ec2Label + "private_ip"
+	ec2LabelSubnetId   = ec2Label + "subnet_id"
 	ec2LabelTag        = ec2Label + "tag_"
+	ec2LabelVpcId      = ec2Label + "vpc_id"
```
Id -> ID

mjibson commented on retrieval/discovery/ec2.go:
```diff
@@ -110,12 +115,25 @@ func (ed *EC2Discovery) refresh() (*config.TargetGroup, error) {
+				if inst.VpcId != nil {
+					labels[ec2LabelVpcId] = model.LabelValue(*inst.VpcId)
+					// take the subnet id of eth0, discard all secondary interfaces
+					labels[ec2LabelSubnetId] = model.LabelValue(*inst.NetworkInterfaces[0].SubnetId)
```
I don't know enough about EC2 to be sure that `*inst` will be non-nil and will have at least one element in its `NetworkInterface` slice. Assuming those are guaranteed when the VPC exists, I suppose this looks correct.

brian-brazil replied:
If there's the potential for an instance to be in multiple subnets, then we should plan for that with the label structure rather than assuming that the user is only interested in one of them.

federicobaldo replied:
Someone may have multiple ENIs, personally I am not sure how to present that information in a usable way. I am assuming that when someone will need that, he/she will have a clear idea of what he needs and how to implement this.

Do you have any idea/guidance on the label structure for this case?

Thanks

brian-brazil replied:
What we've done previously where we have a list is to have a semicolon seperated value, with semicolons at the start/end to make writing regexs against them easier.

I'm not expecting many users to want to filter on these, as subnet ids are fairly opaque.

fabxc replied:
In that case we would concatenate the values into the label, e.g. as done for [Consul tags](http://prometheus.io/docs/operating/configuration/#consul-sd-configurations-consul_sd_config).

federicobaldo replied:
Thanks, will use the same concept of wrapping the values in the separators.

brian-brazil commented on retrieval/discovery/ec2.go:
```diff
@@ -110,12 +115,25 @@ func (ed *EC2Discovery) refresh() (*config.TargetGroup, error) {
 					labels[ec2LabelPublicIP] = model.LabelValue(*inst.PublicIpAddress)
+					labels[ec2LabelPublicDns] = model.LabelValue(*inst.PublicDnsName)
+					// replace address with public dns name if present
+					addr = fmt.Sprintf("%s:%d", *inst.PublicDnsName, ed.port)
```
Please stick to just the private IP address as the default, SD should be kept to something that works in all cases and users can override via relabelling. Having some hosts as public and other as private will cause confusion. In addition this change adds an unnecessary DNS dependency to scraping, and network charges.

federicobaldo replied:
If the user assign a public ip to the instance, he/she may want to use that for connections, as we don't know the user's intention I think that it's fine to leave the prometheus server's location (inside or outside the instance's network) be decisive via dns resolution as dns in a ec2 or vpc network uses split-horizon based on the client's location.

The idea behind that change  is that service discovery should return meaninful information for the client's perspective, so if the Prometheus server is running on the user's laptop, in the ec2 network or in a vpc, the user will not have to change his scrape config as the information will be valid regardless. 

Still I understand your concern about the private ip convention, would a flag in ec2sdconfig [here](https://github.com/federicobaldo/prometheus/blob/ec2_sd_improvements/config/config.go#L681), for instance "use_hostname: true", be ok?

brian-brazil replied:
> Still I understand your concern about the private ip convention, would a flag in ec2sdconfig here, for instance "use_hostname: true", be ok?

Our convention is that anything like that should be handled by relabelling, as there's just too many possible ways a user could want to configure things for us to offer options for all of them. This is something that may be okay in your particular setup, but isn't necessarily correct everywhere (e.g. if I only have a handful of hosts with public IPs in a VPC doing NAT or similar, then this behaviour is incorrect). In general we'd encourage a setup where the prometheus would be within the VPC, using the private IP.

federicobaldo replied:
In that case, without a publicIp the privateIp would be used. Anyway I understand the point, I will drop that and use relabeling. I guess this can be re-evaluated in case other users will have the same scenario.

federicobaldo commented:
In the latest commit, I have addressed the issues you guys pointed out in the comments:
- acronyms all uppercase
- restore privateip as default address label
- manage multiple subnets from multiple ENIs, dealing with possible duplicate
  subnets

Thanks,
Federico

mjibson commented on retrieval/discovery/ec2.go:
```diff
@@ -110,12 +118,34 @@ func (ed *EC2Discovery) refresh() (*config.TargetGroup, error) {
+						subnets = append(subnets, k)
+					}
+					labels[ec2LabelSubnetID] = model.LabelValue(
+						ed.subnetSeparator +
```
This looks wrong: why is the comma both before and after the subnets list? Also, why not assign to multiple labels here instead of joining a bunch? See the tags for an example. (Note: I haven't actually seen what the subnets are, so I'm guessing about this, and could be way wrong.)

fabxc replied:
That's how it is done for the Consul SD tags.
The comma before and after make regexing in relabel configurations significantly simpler.

The tags are spread over multiple labels because they are actually labels, i.e. key/value pairs, so having them in their own label allows actually adressing them. A list on the other hand could only have indexes suffixed to the label names. Here concatenating them with a separator is also simpler in relabel configurations.

mjibson replied:
Ok then. LGTM.

fabxc commented on config/config.go:
```diff
@@ -679,6 +680,7 @@ type EC2SDConfig struct {
 	SecretKey       string   `yaml:"secret_key,omitempty"`
 	RefreshInterval Duration `yaml:"refresh_interval,omitempty"`
 	Port            int      `yaml:"port"`
+	SubnetSeparator string   `yaml:"subnet_separator,omitempty"`
```
As the list items are IDs I think we can just have a fixed separator for them (comma) rather than making it configurable. For Consul tags it was necessary because we have no guarantees about used characters.

federicobaldo replied:
Right. The subnet id can be only in the form "subnet-[a-z0-9]+", so it should be safe to leave comma as a fixed separator. Will wait a few hours in case someone else has some other comments, than I will push a new change removing the SubnetSeparator configuration parameter.

brian-brazil replied:
For the sake of consistency with the default for consul, I think we should stick with semicolon.

We're going to have quite a few of these (e.g. security groups will need similar handling, though those allow semicolons) so consistency will make things simpler for users.

fabxc replied:
Oh yes. Had that wrong in the back of my had.

federicobaldo replied:
I am checking the Consul default tagSeparator [here](https://github.com/prometheus/prometheus/blob/master/config/config.go#L111) and it is a comma, however the default relabeling separator is a semicolon [here](https://github.com/prometheus/prometheus/blob/master/config/config.go#L95). Which one should I pick?

fabxc replied:
Ha, remembered correctly after all – thanks for double-checking.

It was set to comma to be different from relabeling on purpose so you can combine both without ambiguity. Arguably, for a list of similar things a comma is a good fit, while the semicolon is a good fit for relabeling where we concatenate items of different meaning.

So let's stick with the comma.

federicobaldo commented:
- SubnetSeparator option removed from config
- subnetSeparator now is a constant in ec2 sd and match the Consul
  tags default separator

federicobaldo commented:
Hi there, anything else to be done here?

brian-brazil commented on retrieval/discovery/ec2.go:
```diff
@@ -29,10 +31,15 @@ import (
 
 const (
 	ec2Label           = model.MetaLabelPrefix + "ec2_"
+	ec2LabelAZ         = ec2Label + "az"
```
Let's keep this in line with the response we receive, and call this `availability_zone`

federicobaldo commented:
Done, additionally it would be great if we could publish somewhere an example config for this sd.

Here an example for monitoring instances in the same subnet as the prometheus server:
``` yaml
- job_name: ec2_2 
  relabel_configs: 
    - source_labels: ['__meta_ec2_tag_Role'] 
      regex: 'Messaging.*' 
      action: keep
    - source_labels: ['__meta_ec2_tag_Department'] 
      target_label: department 
      regex: '(.*)' 
      replacement: '${1}' 
    - source_labels: ['__meta_ec2_tag_Environment'] 
      target_label: environment 
      regex: '(.*)' 
      replacement: '${1}' 
    - source_labels: ['__meta_ec2_tag_Role'] 
      target_label: role 
      regex: '(.*)' 
      replacement: '${1}' 
    - source_labels: ['__meta_ec2_tag_Name'] 
      target_label: name 
      regex: '(.*)' 
      replacement: '${1}' 
    - source_labels: ['__meta_ec2_availability_zone'] 
      target_label: az 
      regex: '(.*)' 
      replacement: '${1}' 
    - source_labels: ['__meta_ec2_vpc_id'] 
      target_label: vpc 
      regex: '(.*)' 
      replacement: '${1}' 
    - source_labels: ['__meta_ec2_subnet_id'] 
      target_label: subnet 
      regex: '(.*)' 
      replacement: '${1}' 
    - source_labels: ['__meta_ec2_public_dns_name'] 
      target_label: publicdns 
      regex: '(.*)' 
      replacement: '${1}' 
  ec2_sd_configs: 
    - region: us-east-1 
      port: 9100 
```
Here an example on how to monitor instances outside the Prometheus's server network (using the public dns name of the instance):
```yaml
- job_name: ec2_public_dns 
  relabel_configs: 
    - source_labels: ['__meta_ec2_tag_Role'] 
      regex: 'Messaging.*' 
      action: keep
    - source_labels: ['__meta_ec2_public_dns_name']
      target_label: '__address__'
      regex: '(.*)'
      replacement: '$1:9100'
    - source_labels: ['__meta_ec2_tag_Department'] 
      target_label: department 
      regex: '(.*)' 
      replacement: '${1}' 
    - source_labels: ['__meta_ec2_tag_Environment'] 
      target_label: environment 
      regex: '(.*)' 
      replacement: '${1}' 
    - source_labels: ['__meta_ec2_tag_Role'] 
      target_label: role 
      regex: '(.*)' 
      replacement: '${1}' 
    - source_labels: ['__meta_ec2_tag_Name'] 
      target_label: name 
      regex: '(.*)' 
      replacement: '${1}' 
    - source_labels: ['__meta_ec2_availability_zone'] 
      target_label: az 
      regex: '(.*)' 
      replacement: '${1}' 
    - source_labels: ['__meta_ec2_vpc_id'] 
      target_label: vpc 
      regex: '(.*)' 
      replacement: '${1}' 
    - source_labels: ['__meta_ec2_subnet_id'] 
      target_label: subnet 
      regex: '(.*)' 
      replacement: '${1}' 
    - source_labels: ['__meta_ec2_public_dns_name'] 
      target_label: publicdns 
      regex: '(.*)' 
      replacement: '${1}' 
  ec2_sd_configs: 
    - region: us-east-1 
      port: 9100 
```
The only awkward thing with the latest example is that the "ec2_sd_config: port" value is discarded in the process and must be explicitly written in the relabel config. Do you know any better way to do this?

brian-brazil commented:
I think you might be over doing the labels a tad. With that many you might want to consider the `labelmap` action.

> The only awkward thing with the latest example is that the "ec2_sd_config: port" value is discarded in the process and must be explicitly written in the relabel config. Do you know any better way to do this?

The port is intended as a default for the cases where you aren't doing relabelling. You could just exclude it.

federicobaldo commented:
Thanks for the clarification,  will try that as it could really simplify the config. In the meantime please let me know if there's anything else to do on this one in order to get this into master.

brian-brazil commented:
Can you squash your commits? @fabxc will be back tomorrow and give it a final look.

federicobaldo commented:
Done! Thanks

fabxc commented:
Thanks :+1: 

Example configurations would indeed be very helpful. Feel free to file another PR with those in the `documentation/examples` directory :)
